hedera-node/docs/proposals/block-stream-bucket-uploader.md

```diff
@@ -0,0 +1,398 @@
+- Exponential backoff with random jitter
+- Exception handling and propagation
+- Operation timeout management
+- Detailed logging of retry attempts
+- Support for operations that can throw checked exceptions
+
+
+### BucketUploadListener Interface
+Interface for components that need to be notified when block files are closed and ready for upload.
+
+Key Operations:
+- `onBlockClosed(Path blockPath)` - Called when a block file is closed and ready for upload
+
+#### FileBlockItemWriter Updates
+The FileBlockItemWriter class needs to be enhanced to support notifying listeners when blocks are closed.
+
+Required Changes:
+- Add list of registered BucketUploadListener instances
+- Add methods to register/unregister listeners
+- Update closeBlock() to notify listeners after successful block closure
+- Handle errors during block closure and notify listeners
+
+### BucketUploadManager
+
+The BucketUploadManager is responsible for coordinating block uploads to multiple cloud storage providers. It listens for block closure events and manages the asynchronous upload process, including directory management and metrics collection.
+
+Key Implementation Details:
+- Implements BucketUploadListener interface to receive block closure notifications
+- Manages concurrent uploads to multiple cloud providers using an ExecutorService
+- Manages block files by moving them to appropriate directories based on upload status:
+  - Successful uploads: moved to `/uploaded` directory
+  - Hash mismatches: moved to `/hashmismatch` directory
+  - Failed uploads: remain in original directory for retry
+- Performs hash validation to detect corrupted or incomplete uploads
+- Collects detailed metrics on upload operations including:
+  - Success/failure counts per provider
+  - Upload latency measurements
+  - Hash mismatch detection
+  - Skipped upload tracking
+- Handles upload failures with proper error logging and metrics
+- Uses CompletableFuture for non-blocking asynchronous operations
+
+
+### BlockReuploadManager
+The BlockReuploadManager is responsible for initiating re-uploads of failed block uploads after node restarts or re-uploading blocks that were attempted to be uploaded more than the configured retry limit.
+
+##### Key Responsibilities
+- Managing retry attempts for failed uploads while respecting configured retry limits
+- Coordinating with the BucketUploadManager to re-attempt uploads
+- Handling error scenarios gracefully with appropriate logging and failure reporting
+
+##### Startup Operations
+- Scan block directory on node startup
+- Identify blocks that need to be uploaded (files in the current directory)
+- Build recovery queue
+
+##### Recovery Processing
+- Initiate recovery of incomplete uploads
+- Track recovery progress
+- Report recovery metrics
+
+### BlockRetentionManager
+
+The BlockRetentionManager is responsible for managing the lifecycle of block files in the `/uploaded` directory after a configurable retention period.
```

Comment:
Should this also be responsible for rentention or cleanup of the files that have a hash mismatch (as there will be many of these and they will be most likely just due to another node having same file uploaded first) ?